---
title: "Parasitic genera in WWTPs based on both metagenomic and metatranscriptomic data"  
author: "Jule Freudenthal"  
date: "`r Sys.Date()`"  
output: github_document
always_allow_html: yes
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set global working directory
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

**R version:** 3.6.2 (2019-12-12), Dark and Stormy Night 
**Packages**

* kableExtra v. 1.3.4  
* rlist v. 0.4.6.1  

```{r, message=FALSE, warning=FALSE, results='hide', echo=FALSE}
# Install packages
if (!require("kableExtra")) install.packages("kableExtra")
if (!require("rlist")) install.packages("rlist")
```
```{r, message=FALSE, warning=FALSE}
# Load packages
library(kableExtra)
library(rlist)
```

We assign functional traits to the taxa identified, using published reference databases. Based to the these trait databases, we classified the following taxa as **parasites**:  

1. **Protists**: All protist genera associated with human and animal gut and/or feces.  
2. **Prokaryotes, fungi, microscopic metazoa**: All prokaryote, fungal and microscopic-metazoan genera that include potentially pathogenic species to humans and animals.    

Now, we'll give an overview of all identified parasitic genera.

```{r, message=FALSE, warning=FALSE}
# Load table with names of preprocessed count tables 
filenames_count_tables <- read.table("./R/Lists/Preprocessing_data.csv", header = TRUE, 
                                     row.names = 1, sep = ";", colClasses = "character")

# Create open list for tables
parasitic_taxa <- list()

# Loop iterates over the data types (DNA/RNA data)
for(data_type in c("DNA", "RNA")){
  
  # Load count table and metadata
  count_table <- read.table(paste0("./DataToAnalyse/PreprocessedData/", data_type, "_" , 
                                   filenames_count_tables[9,]), 
                            header = TRUE, sep = ";", dec = ".", row.names = 1)
  
  # Specify the microbial community
  count_table$`Microbial community` <- "Prokaryotes"
  count_table[count_table$Database == "PR2",]$`Microbial community` <- "Protists"
  count_table[count_table$Database == "PR2" & 
                count_table$Phylum == "Metazoa",]$`Microbial community` <- "Metazoa" 
  count_table[count_table$Database == "PR2" & 
                count_table$Phylum == "Fungi",]$`Microbial community` <- "Fungi"
  
  # Subset data set, keep only parasitic genera
  parasites <- count_table[count_table$Traits == "parasite", c("Genus", "Microbial community")]
  
  # Save the table in the list
  parasitic_taxa <- list.append(parasitic_taxa, parasites)
}

# Merge both tables to one
parasitic_taxa <- do.call(rbind, parasitic_taxa)

# Unique gut-associated genera
parasitic_taxa <- unique(parasitic_taxa)

# Order table
parasitic_taxa <- parasitic_taxa[order(parasitic_taxa$`Microbial community`, 
                                       parasitic_taxa$Genus),]

# Genus names italic
parasitic_taxa$Genus <- cell_spec(parasitic_taxa$Genus,italic = T)

# Subset table
parasitic_taxa <- cbind(parasitic_taxa[seq(nrow(parasitic_taxa)/2),
                                       c("Genus", "Microbial community")],
                        parasitic_taxa[((nrow(parasitic_taxa)/2)+1):nrow(parasitic_taxa),
                                       c("Genus", "Microbial community")])

# Delete row names
rownames(parasitic_taxa) <- NULL

# Visualize table
kable(parasitic_taxa, escape = F, booktabs = TRUE, align = "lclc") %>%
  kable_classic(full_width = F, html_font = "Calibri")  %>%
  kable_styling(font_size = 12, bootstrap_options = "condensed", position = "center", 
                latex_options = "scale_down") 
```