---
title: "Microbial community structure and environmental factors in the separate WWTP compartments"  
author: "Jule Freudenthal"  
date: "`r Sys.Date()`"  
output: github_document
always_allow_html: yes
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Set global working directory
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

**R version:** 3.6.2 (2019-12-12), Dark and Stormy Night  
**Packages**

* docstring v. 1.0.0 
* ggplot2 v. 3.3.5  
* ggpubr v. 0.4.0  
* ggrepel v. 0.9.1  
* mgsub v. 1.7.2  
* rlist v. 0.4.6.1  
* zeallot v. 0.1.0  

```{r, message=FALSE, warning=FALSE, results='hide', echo=FALSE}
# Install packages
if (!require("docstring")) install.packages("docstring")
if (!require("ggplot2")) install.packages("ggplot2")
if (!require("ggpubr")) install.packages("ggpubr")
if (!require("ggrepel")) install.packages("ggrepel")
if (!require("mgsub")) install.packages("mgsub")
if (!require("rlist")) install.packages("rlist")
if (!require("zeallot")) install.packages("zeallot")
```
```{r message=FALSE, warning=FALSE}
# Load packages
library(docstring)
library(ggplot2)
library(ggpubr)
library(ggrepel)
library(mgsub)
library(rlist)
library(zeallot)

# Load functions
source("./Functions/calculate.NMDS.R")
source("./Functions/ordination.metadata.R")
source("./Functions/plot.NMDS.R")
source("./Functions/style.sheet.R")
source("./Functions/style.text.R")
```

To evaluate the influence of environmental factors on the WWTP microbial community, we compute biplots based on Bray-Curtis dissimilarities for the individual compartment types, for both rDNA and rRNA relative abundance data (function metaMDS, package vegan). We fit the environmental data measured by [Ju et al. (2019)](https://doi.org/10.1038/s41396-018-0277-8) (pH, dissolved oxygen, dissolved organic carbon, total nitrogen, total phosphorus, temperature, and hydraulic retention time) onto the ordinations using envfit (vegan). We will correct the resulting p-values for multiple testing according to Benjamini & Hochberg. Next, we scale (multiply) significant environmental vectors with their correlation value and add to the NMDS plots. 

```{r, warning=FALSE, results='hide', message=FALSE, fig.width=10, fig.height=5, fig.align='center', fig.show='hold'}
# Create vectors specifying the colors per sample place
colors_sample_places <- c("goldenrod1", "firebrick2", "forestgreen", "blue")
names(colors_sample_places) <- c("INF","DNF","NFC","EFF")

# Create open list
NMDS_plots <- list()
  
# Loop iterates over the data types (DNA/RNA data)
for(data_type in c("DNA", "RNA")){
  
  # Create open list
  plots <- list()

  # Load count- and meta- and taxonomy data
  load(paste0("./DataToAnalyse/RData/", data_type, "_data.rds"))
  
  # Check if column names of the count data and row names of the metadata match
  if(!identical(colnames(counts), rownames(metadata))){
    stop("Column names of the count data and row names of the metadata match")
  }

  # Normalize counts (relative counts)
  normalized_counts <- sweep(x = counts, MARGIN = 2, STATS = colSums(counts), FUN = '/')
  
  # Loop iterates over the WWTP compartments 
  for(compartment in c("INF", "DNF", "NFC", "EFF")){
    
    # Subset data, keep only samples belonging to one compartment
    normalized_counts_comp <- normalized_counts[,grepl(compartment, colnames(normalized_counts))]
    metadata_comp <- metadata[grepl(compartment, rownames(metadata)),]
    
    # Delete 'null' taxa, taxa that do not occur in this compartment
    normalized_counts_comp <- normalized_counts_comp[rowSums(normalized_counts_comp) != 0,]

    #----------------------------- Calculate NMDS -----------------------------#
    # Calculate ordination
    NMDS <- calculate.NMDS(normalized_counts_comp) 
  
    # Visualization
    plot <- plot.NMDS(NMDS$Eigenvectors, group.attrib = metadata_comp$Sample_Place, 
                      point.size = 2, draw.ellipse.by = "Group", 
                      cluster.attrib = metadata_comp$Location)
  
    # Adjust plot
    if(data_type == "DNA"){
      plot <- style.sheet(plot, x.axis.label = "", plot.title = compartment)
    } else {
      plot <- style.sheet(plot, plot.title = "")
    }
    if(compartment != "INF"){plot <- style.sheet(plot, y.axis.label = "")}
    plot <- style.sheet(plot, colors = colors_sample_places[compartment], 
                        fill.colors = colors_sample_places[compartment], 
                        fill.colors.opacity = 0.2, legend.title.by.shape = "WWTP locations",
                        legend.position = "bottom", show.legend.by.color = FALSE, 
                        show.legend.by.fill = FALSE)
    
    # Add stress value
    plot <- plot + annotate("text", x=Inf,y=-Inf,hjust=1.1,vjust=-0.5,
                            label=paste("Stress value:", round(NMDS$StressValue,3)))
    
    #------------------------- Fit environmental data -------------------------#
    # Fit environmental data onto ordination
    c(environmental.vectors, environmental.factor) %<-% 
      ordination.metadata(metadata_comp, ordination = NMDS$Ordination)
  
    # Keep only environmental vectors with p value < 0.05
    environmental.vectors <- environmental.vectors[environmental.vectors$p.values < 0.05,]
  
    # Check for significant environmental vectors
    if(nrow(environmental.vectors) != 0){
      # Change names of environmental factors
      environmental.vectors$Labels <- mgsub(row.names(environmental.vectors), 
                                            c("mg_per_L", "_C", "hours", "_"),
                                            c("[mg/L]", " [C]", "[h]", " "))
      
      # Format text, add stars to indicate significance
      environmental.vectors$Labels <- 
        style.text(names = environmental.vectors$Labels, p.values = environmental.vectors$p.values,
                   italic.text=rep(FALSE, nrow(environmental.vectors)))
      
      # Check if xlims and ylims are in range or if arrows are out of rang
      if(!all(environmental.vectors$x1 > layer_scales(plot)$x$range$range[1] & 
              environmental.vectors$x1 < layer_scales(plot)$x$range$range[2])){
        plot <- style.sheet(plot, xlims = c(min(c(environmental.vectors$x1, 
                                                  layer_scales(plot)$x$range$range))*1.05, 
                                            max(c(environmental.vectors$x1, 
                                                  layer_scales(plot)$x$range$range))*1.05))
      }
      if(!all(environmental.vectors$y1 > layer_scales(plot)$y$range$range[1] & 
              environmental.vectors$y1 < layer_scales(plot)$y$range$range[2])){
        plot <- style.sheet(plot, ylims = c(min(c(environmental.vectors$y1, 
                                                  layer_scales(plot)$y$range$range))*1.05, 
                                            max(c(environmental.vectors$y1, 
                                                  layer_scales(plot)$y$range$range))*1.05))
      }
      
      # Add arrows
      plot <- plot + geom_segment(data=environmental.vectors, aes(x=x0, y=y0, xend=x1, yend=y1),
                                  inherit.aes = F, arrow=arrow(), color="black")
      plot <- plot + geom_text_repel(data=environmental.vectors, aes(x=x1, y=y1,label=Labels),
                                     inherit.aes = F, size=4, parse = T, force = T)
      
    }
    
    # Format all y-axis label into scientific notation
    plot <- plot + scale_y_continuous(labels = function(x) formatC(x, format = "e", digits = 1))
    
    # Adjust space around plot
    ifelse(data_type == "DNA", plot <- plot + theme(plot.margin = margin(t = 2, r = 2, b = -5, l = 2)), 
           plot <- plot + theme(plot.margin = margin(t = -5, r = 2, b = 2, l = 2)))
      
    # Save plot in list
    plots <- list.append(plots , plot)
  }
  # Get legend
  if(data_type == "DNA"){legend <- get_legend(plots[[3]])}
  
  # Arrange plots
  plots_arranged <- ggarrange(plotlist=plots, ncol = 4, nrow=1, common.legend = TRUE, legend = "none")
  plots_arranged <- annotate_figure(plots_arranged,  left = text_grob(
    ifelse(data_type=="DNA","Metagenomics", "Metatranscriptomics"), rot = 90, size = 16, face = "bold"))
  
  # Save plot in list
  NMDS_plots <- list.append(NMDS_plots , plots_arranged)
  
}

# Arrange plots
NMDS_plots <- ggarrange(plotlist=NMDS_plots, ncol = 1, nrow=2)
NMDS_plots <- ggarrange(plotlist=list(NMDS_plots, legend), 
                        ncol = 1, nrow = 2, heights = c(1, 0.2))
print(NMDS_plots)
```

NMDS biplots show the microbial community composition in association with environmental data for each WWTP compartment. Metagenomic and metatranscriptomic data are shown separately. The distribution of the samples (symbol-coded by WWTP location) is visualized by the ellipses. Significant environmental vectors are shown as arrows (* p < 0.05; ** p < 0.01; *** p < 0.001). Environmental vectors: TN = total nitrogen.

This analysis shows that environmental factors such as pH, total phosphorus and dissolved organic matter do not correlated significantly with the diversity of the microbial community at the individual compartment level (with one exception).

```{r, echo=FALSE, results='hide'}
# Create new directory for plot
if(!file.exists("./Results")){dir.create("./Results")}
if(!file.exists("./Results/NMDS")){dir.create("./Results/NMDS")}

# Save plot as pdf
pdf("./Results/NMDS/NMDS_CompartmentsWWTP.pdf", width = 10, height = 5)
print(NMDS_plots)
dev.off()
```
